오픈 마켓2 [STEP1] 허황, 애플사이다

just1103 opened the pull request:
메이슨 안녕하세요. @myssun0325
애플사이다 @just1103, 허황 @hwangjeha 입니다.
STEP1 완료하여 PR 드립니다. ViewController가 비대해져서 DataSource 타입을 생성하여 구분해줬는데, 부족한 점이 많을 것 같습니다. 😅 이번 리뷰도 잘 부탁드립니다. :)

## STEP1 구현내용
### 구현 화면 

|이미지 추가|이미지 수정/삭제|상품 등록 실패|상품 등록 성공|키보드 입력 처리|
|-|-|-|-|-|
|![](https://i.imgur.com/oDjsZYI.gif)|![](https://i.imgur.com/rr6Dd1h.gif)|![](https://i.imgur.com/edftfPN.gif)|![](https://i.imgur.com/GDTV2wS.gif)|![](https://i.imgur.com/cMXZ1n6.gif)|

### 1. HTTP Post
Gettable, Postable 프로토콜을 만들고 기존의 APIProtocol를 상속받았습니다.
Gettable, Postable을 분리한 이유는 이번 HTTP Post를 구현하면서 Post에 필요한 contentType, body 등을 APIProtocol의 프로퍼티로 넣게되면 get을 요청하는 API구조체들에도 프로퍼티가 생기기 때문이었습니다.

HTTP Post 메서드에 필요한 body를 만드는 MultipartFormData 타입을 구현했습니다.
오픈마켓 서버 명세의 고정된 body("params", "images")의 키를 사용하는 방식이 아니라 유동적으로 body를 구성할 수 있도록 구현했습니다.

### 2. 상품 등록/수정 화면의 구성
`상품 등록 화면` 및 `상품 수정 화면`의 차이점이 `이미지 추가/수정 기능` 가능 여부 밖에 없기 때문에 두 화면을 구성할 때, 동일한 ViewController 및 동일한 Scroll View를 공유하도록 설계했습니다.
(STEP1에서 `상품 등록 화면` 및 `상품 수정 화면`을 구현해야 했는데, `상품 수정화면`을 아직 구현하지 않았습니다. 사용자가 `상품 목록 화면`에서 상품을 탭하여 `상품 상세화면`을 확인할 때, vendor id를 비교하여 해당 상품의 vendor일 경우 `상품 수정 화면`으로 넘어갈 수 있는 버튼을 추가하기로 설계했습니다. STEP2에서 구현할 예정입니다.)

화면의 UI를 구성할 때는 `Scroll View` 내부에 `Vertical StackView`를 올리고, TextFeild, TextView 등을 올려서 구현했습니다. 
또한 제품 상세내용을 작성하는 `TextView`에도 `Placeholder`를 추가하여 사용자가 쉽게 글자수 제한 등을 알 수 있게 했습니다.

### 3. 상품 등록 화면의 이미지 추가 버튼 및 자동 Scroll 기능 구현
이미지를 수평 방향으로 Scroll할 수 있게 구현하기 위해 `CollectionView`를 사용했습니다. 그리고 사용자가 상품 이미지를 등록할 수 있도록 `CollectionView의 Footer`를 만들고, `이미지 추가 버튼`을 생성했습니다.
또한 이미지를 추가한 직후에 사용자가 추가버튼을 바로 볼 수 있도록 CollectionView를 자동으로 Scroll하게 구현했습니다. 이 과정에서 데이터를 reload한 이후에 Scroll할 수 있도록 completionHandler 기능을 넣은 `reloadDataCompletion` 메서드를 구현했습니다.

### 4. 키보드가 TextView를 가리지 않도록 수정
Keyboard 관련 Notification을 통해 키보드가 나타날 때 ScrollView의 `Bottom Content Inset`을 `키보드 높이`만큼 지정하도록 했습니다. 또한 키보드가 내려할 때 다시 Inset을 없애도록 했습니다.

### 5. Collection View Cell FirstResponder
키보드를 구현하면서 키보드 영역이 아닌 다른 영역을 눌렀을 때 키보드가 내려갈 수 있도록 ViewController의 view에 UITabGesture를 추가해줬습니다.
여기서 문제는 image Collection View의 Cell을 눌렀을 때 Collection View의 델리게이트 메서드 `didSeleteItemAt` 메서드가 호출되지 않는 문제가 발생했습니다.
저희가 원인을 유추해봤는데 cell에는 탭 제스쳐를 인식할 수 없어 탭 제스쳐가 ViewController의 View까지 도달했다는 생각을 했습니다.
따라서, 저희는 Cell위에 버튼을 만들고 그 버튼을 눌렀을 때 어떤 Cell이 눌렸는지 판단하는 기능을 추가해서 문제를 해결했습니다.
```swift
// 상위 뷰 중 매개변수로 받은 ofType 타입이 있다면 그 타입을 반환해주는 메서드
private extension UIView {
    func findSuperview<T>(ofType: T.Type) -> T? {
        var currentView = self
        var resultView: T?
        while let currentSuperview = currentView.superview {
            if currentSuperview is T {
                resultView = currentSuperview as? T
                break
            }
            currentView = currentSuperview
        }
        return resultView
    }
}
// ...
 @objc func editImageOfSelectedItem(_ sender: UIButton) {
     // Cell의 Button을 눌렀을 때 그 버튼의 superview 중 cell 타입이 있다면 
     // cell의 indexPath를 받아와서 이미지를 수정/삭제 할 수 있다.
        guard let productImageCell = sender.findSuperview(ofType: ProductImageCell.self),
              let indexPath = productImageCell.indexPath else {
            return
        }
     /...
 }
```

### 6. 메모리 캐시(보너스 스탭)
저번 오픈마켓1에서 캐시를 구현해보라는 보너스 스탭이 있어 NSCache를 사용해서 `메모리 캐시`를 구현했습니다.
저희는 서버에서 받아오는 이미지의 URL을 캐시에 key로 사용하고, 이미지를 로드하는 흐름을 설계해봤습니다.
1. key(이미지의 URL)가 메모리에 있는지 확인한다.
2. 메모리에 이미지가 있다면 메모리에서 꺼내서 로드한다.
3. 메모리에 이미지가 없다면 이미지의 URL로 데이터를 다운받아 변환한 이미지를 생성한다.
4. 생성한 메모리를 캐시에 저장한다.
위와 같은 흐름으로 이미지가 재사용될 때마다 서버에서 이미지를 다운받는 방식이 아닌 메모리에 이미지가 있는 경우 캐시를 사용하는 방법으로 이미지 캐시를 구현했습니다.

## 고민한 점 및 궁금한 점
### 1. Tap Gesture
사용자가 화면에서 `이미지 CollectionView` 또는 `textField/textView`가 아닌 영역을 탭하면, 키보드가 내려가도록 구현하기 위해 ViewController에 `Tap Gesture Recognizer`를 추가했습니다.
하지만 이로 인해 사용자가 등록한 이미지를 다시 수정하기 위해 `이미지 CollectionView`의 이미지를 탭했을 때, ViewController가 제스쳐를 가로채서 CollectionView가 제스쳐를 인식하지 못하는 문제가 발생했습니다.
따라서 CollectionView의 Cell 내부에 Button을 추가하고, 해당 버튼에 이미지를 넣도록 하여 해결했습니다. (버튼은 항상 Tap Gesture의 first responder이기 때문입니다.) 현업에서는 이 문제에 대해 어떤 방법을 사용하는지 궁금합니다.

### 2. ViewController 거대화
기능을 하나씩 구현하면서 ViewController이 점점 커지는 문제(?)발생했습니다.
따라서, ViewController의 Extension을 별도의 파일로 만들어서 ViewController의 소스코드를 분리시켰습니다.
사실 저희가 별도의 파일로 만든건 ViewController의 Extension이지 ViewController의 기능을 분리시켜준건 아니라고 생각합니다.
ViewController이 거대해질 경우 Extension을 사용해서 다른 파일로 분리해줘도 될까? 라는 고민을 했습니다.

또, 동일한 파일 내 ViewController를 확장한 경우 private 메서드, 프로퍼티에 접근이 가능하지만 다른 파일로 Extension해준 경우 private 메서드, 프로퍼티에 접근을 하지 못하는 경우가 발생했습니다.
다른 파일이더라도 extention이므로 "구현부"에 속하는데 왜 다른 파일에 만들어준 ViewController의 extension은 private에 접근을 못하는 것인지도 궁금합니다.

### 3. 등록할 이미지 크기 축소
이미지 크기를 300KB로 제한하라는 요구사항이 있어 이미지 크기를 줄이는 기능을 추가했습니다.

처음 이미지 사이즈를 줄이는 방법 메서드(preparingThumbnail)를 찾았지만 iOS 15부터 제공하는 기능이라서 `#available(iOS 15.0, *)`을 사용해서 버전 별로 기능을 분기처리했습니다.

이미지 크기를 줄이는 로직은 
`비율 = 제한된 용량(300KB) / imagePicker에서 받아온 이미지의 데이터 크기`
하게 되면 제한된 용량 대비 현재 이미지 용량의 비율이 나오는데
위에서 구한 비율로 이미지의 width, height 크기를 재조정했습니다.

### 4. CollectionView의 deleteItem 메서드
사용자가 이미지 우상단의 `x 버튼`을 탭하여 등록한 이미지를 삭제할 경우, 아래의 코드를 통해 `CollectionView` 및 `productImages 배열`에 반영하도록 했습니다. 하지만 앱이 중지되지는 않았지만 console log에서 `업데이트 이후 item 개수 = 업데이트 이전 item 개수 - 삭제한 item 개수`가 되지 않았다는 경고가 나타났습니다.
따라서 deleteItems 메서드를 reloadData 메서드로 대체했습니다. reloadData를 호출하면 DataSource 메서드인 cellForItemAt가 호출되기 때문에 이를 활용했습니다. 이 방법이 적절한지, 그렇다면 deleteItems 메서드는 어떤 경우에 사용하는지 궁금합니다. (다른 예제를 찾아봐도 이번 상황과 비슷한 경우에 사용하고 있었습니다.)

```swift 
@objc func removeImage(_ sender: UIButton) {
    guard let productImageCell = sender.findSuperview(ofType: ProductImageCell.self),
          let indexPath = productImageCell.indexPath else {
        return
    }
    
    // 수정 이전
    imageCollectionView.deleteItems(at: [indexPath])  // delete 직후에 reload됨
    productImages.remove(at: indexPath.item)
    
    // 수정 이후
    productImages.remove(at: indexPath.item)
    imageCollectionView.reloadData()  // dataSource 메서드 (cellForItemAt)가 호출되므로 해당 indexPath의 item이 삭제됨
}
```

### 5. Custom Cell, ProductManagementScrollView 등 View 구현내용
오픈마켓1 PR에서 Custom Cell 관련한 부분은 별도의 피드백을 주시지 않았는데, 저희가 제대로 구현한 것이 맞을지 궁금합니다. 😅 (인터페이스 빌더를 사용하지 않고 코드로만 View를 구현해봤는데, 개선할 점이 많을 것 같아서 찔렸어요..) 피드백 부탁드립니다. :)

just1103 committed:
refactor: ViewTypeSegmentedControl -> LayoutKindSegmentedControl 타입명 변경 #4

- OpenMarketViewController의 불필요한 self 삭제
- OpenMarketViewController의 Mark 주석 추가

Commit:
refactor: Strikable 프로토콜 추가 및 StrikableUILabel 생성 #4

- String+Extension 제거
- CustomCell priceLabel 타입 변경 (UILabel -> StrikableUILabel)

just1103 committed:
refactor: Strikable 프로토콜 및 Strikable 타입 삭제 #4

- UILabel 익스텐션으로 변경함

Commit:
fix: CollectionView 모양 변경시 잔상버그 개선 #4

- ProductsCollectionView 타입 추가
- reloadDataCompletion 메서드 추가
- fadeIn, fadeOut 메서드 추가
- toggleViewTypeSegmentedControl 메서드 개선(함수 분리)

just1103 committed:
docs: STEP1 README 업데이트

Commit:
docs: README.md STEP 2 Update

just1103 committed:
fix: Layout Conflicts 수정 #4

- Inset을 반영하여 Item Layout의 width 수정

Commit:
feat: AddProductViewController NavigationBar 구성 #6

- AddproductViewController title 추가
- navigationBar Button 추가(취소, 완료)

just1103 committed:
feat: ProductManagementView 타입 추가 #6

- 상품 등록/수정 화면을 위한 Custom View 타입 추가
- Currency 타입 CustomStringConvertible 프로토콜 채택
 <타입>: <제목>

Commit:
refactor: ProductManagementView 코드 개선 #6

- RoundedRectTextField 타입 생성(borderStyle = .roundedRect)
- UITextField -> RoundedRectTextField 타입 변경

just1103 committed:
feat: descriptionTextView에 placeholder 추가 #6

Commit:
feat: ProductImageCell 추가 #6

just1103 committed:
feat: AddProductImageFooterView 타입 추가 #6

- imageCollectionView의 Footer 추가
- AddProductImageActionSheetText 타입 추가
- Footer의 addButton을 탭하면 ActionSheet를 present하는 기능 추가

Commit:
feat: ProductImage 수정/삭제 기능 추가 #6

just1103 committed:
feat: imageCollectionView 최우측으로 자동 Scroll 기능 추가 #6

- Action Sheet의 Alert Title을 이미지 등록/수정 기능에 따라 변경하도록 수정
- UIView 익스텐션하여 findSuperview 메서드 추가
- ProductImageCollectionView 타입 및 reloadDataCompletion 메서드 추가

Commit:
feat: MultipartFormData 타입 생성 #6

- ContentDisposition 열거형 생성
- EncodingCharacters 열거형 생성
- BoundaryGenerator 열거형 생성
- Data Extension append 메서드 추가

just1103 committed:
feat: createImageFormData 메서드 추가 #6

- ImageContentType 열거형 추가
- Gettable 및 Postable 프로토콜 추가
- URLRequest 익스텐션에서 Postable 프로토콜에 대한 처리 추가

Commit:
fix: 상품 등록 Post 기능 버그 수정 #6

just1103 committed:
feat: 업로드할 이미지 크기 조정 기능 추가 #6

- UIImage 익스텐션 추가
- AddProductViewController의 postProductRegister 메서드 수정

Commit:
feat: 사용자 입력을 받아서 Post하는 기능 #6

- 상품등록 성송 alert 추가

just1103 committed:
feat: 사용자 입력값을 검증하는 기능 추가 #6

- UserInputChecker 타입 추가
- GenerateUserInputError 열거형 추가
- createUserInputData 메서드 반환타입을 Result로 변경
- validateUserInput 메서드 추가

Commit:
feat: 이미지 5장이상 못 올리게 처리 #6

just1103 committed:
feat: 키보드가 textView를 가리지 않도록 하는 기능 추가 #6

- addObserverKeyboardNotification 메서드 추가
- tapBehindViewToEndEdit 메서드 추가

Commit:
refactor: 화면별 파일 분리 #6

just1103 committed:
chore: 줄바꿈 및 네이밍 수정 #6

Commit:
chore: productManagementView 프로퍼티명 변경 #6

- productManagementView -> productManagementScrollView

just1103 committed:
refactor: ViewTitle 열거형 추가 #6

Commit:
feat: 상품 리스트 갱신 기능 추가 #6

- ProductListStackView 추가(새로고침 버튼, 컬렉션 뷰)
- productCollectionView RefreshControl 추가
- 상품 새로고침 기능 추가
- 주기적으로 서버의 상품 정보를 가져오는 기능 추가

just1103 committed:
refactor: 상품명 및 재고수량의 CompressionResistant Priority 수정 #6

- 상품명 Label이 재고수량 Label 길이에 묻히지 않도록 수정
- 단, 품절 시 상품명 Label보다 재고수량 Label이 보이도록 함

Commit:
feat: 상품 목록 화면 회전 고정(Portrait) #6

just1103 committed:
fix: 화면전환 시 상품등록 화면의 Layout 변경 기능 추가 및 버그 수정 #6

- ScrollView의 widthAnchor 수정
- imageAddFooterViewSize 수정

Commit:
feat: 상품 목록 새로고침 버튼 누르면 스크롤 최상단 기능 추가 #6

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,148 @@
+}
+
+// MARK: - SegmentControl
+extension OpenMarketViewController {
```
고민에서도 남기셨지만 분리를 하거나 뷰컨에 두거나 하는 방법보다 이렇게 했을 경우 어떤 장점이 있다고 생각하시나요~?

hwangjeha replied:
물리적으로 소스코드가 분리되서 관련된 메서드를 따로 분리해줄 수 있지만, 실제 뷰컨을 분리한게 아니기 때문에 접근 제어 등 단점이 더 많은것 같습니다.
델리게이트 패턴을 적용해서 데이터 소스, 뷰컨 각자 역할에 맞게 분리해서 개선했습니다😊

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,148 @@
+        case .list:
+            let inset: Double = 10
+            let listCellSize: (width: CGFloat, height: CGFloat) = (view.frame.width - inset * 2,
+                                                                   view.frame.height * 0.077)
```
여기에 표현되는 0.077이라는 값이 무엇인지 유추하기 어려운거 같습니다. 
의미있는 코드다 되도록 수정을 하거나 코멘트로 적어두면 좋을듯해용 :)

hwangjeha replied:
리스트 형태의 컬렉션 뷰일 때 셀이 화면에 13개 정도 보여지게끔 1을 13으로 나눈 값입니다.
화면 대비 셀의 비율을 의미하는 상수 cellHeightRatio을 만들어서 네이밍으로 의미를 전할하는게 의미있는 코드가 될 수 있을 것 같습니다:)

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,148 @@
+    }
+    
+    var currentLayoutKind: LayoutKind = .list
+    var products: [Product]?
```
이런식으로 데이터를 외부에서도 수정할 수 있도록 하면 어떤점에서 좋고 어떤점이 안좋을까요?
외부에서 필요한 값에 대해서만 줄 수 있도록 하고, data로 부터 받아온 값들도 지금은 뷰컨트롤러가 일을해서 직접 넣어주고 있는데, 그보다는 데이터소스한테 데이터를 넘겨주면 필요한값들을 알아서 세팅할 수 있도록 해주는건 어떤가요?

just1103 replied:
코드의 다른 부분에서 임의로 products에 접근하여 수정할 수 있기 때문에 안전하지 않은 것 같습니다.
위 부분과 마찬가지로 델리게이트 패턴을 적용해서 데이터 소스, 뷰컨 각자 역할에 맞게 분리해서 개선했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        return .portrait
+    }
+    
+    var dataSource = OpenMarketDataSource()
```
여기서 dataSource를 열어둔 이유가 OpenMarketViewController의 extension에서 쓰기 때문인데,  개인적으로 조금 어색한거 같습니다. 결국은 OpenMarketViewController안에서만 쓰는 속성인데 파일의 구조를 위해 코드적으로 이것을 열어야한다면 여기서 취할 수 있는 이점은 무엇일까요?

just1103 replied:
위와 동일하게 개선했습니다!

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    private var segmentedControl: LayoutKindSegmentedControl!
+    private(set) var productCollectionView: ProductsCollectionView!
+    private(set) var productListStackView = ProductListStackView()
+    private var activityIndicator: UIActivityIndicatorView!
```
굳이 여기가 옵셔널로 처리되어야하는 이유가 있을까요? 이 외에도 segmentedControl도 그냥 생성하지 않은 이유가 궁금합니다~

just1103 replied:
사실 `IBOutlet 변수`를 보통 `옵셔널 타입`으로 선언했던 것을 봐왔기 때문에 `옵셔널 타입`이 일반적인 방법이라고 생각했습니다. 그리고 변수를 선언할 때 `기본값`을 지정하더라도 나중에 각각의 View를 setup하는 메서드에서 어차피 변수 관련된 설정을 다시 해야하기 때문에, 일을 두 번 한다는 느낌을 받아서 `옵셔널`을 사용했습니다.

다시 생각해보니 기존에 `옵셔널`을 사용했던 것은 인터페이스 빌더의 UI 요소와 IBOutlet 변수를 연결해야 하고, 연결이 끊어질 가능성이 존재했기 때문에 불가피하게 `옵셔널`을 사용했던 것 같습니다. 현재 코드에서는 인터페이스 빌더를 사용하지 않기 때문에 굳이 `옵셔널`을 사용할 필요가 없을 것 같습니다. 
이렇게 `옵셔널`을 사용하는 것보다 `기본값`을 지정하는 게 더 안전하고 효율적인 방법이라고 판단해도 괜찮을까요?

* `segmentedControl`의 init을 수정하여 `기본값`을 지정하도록 개선했습니다.
* 덧붙여 `productCollectionView`는 옵셔널로 처리한 이유는 `CollectionView`의 delegate 및 dataSource를 해당 `ViewController`로 할당해야 하기 때문에 `setupCollectionView` 메서드에서 `productListStackView` 내부의 `productCollectionView`의 참조를 할당하도록 하기 위함입니다.

myssun0325 replied:
'이렇게 옵셔널을 사용하는 것보다 기본값을 지정하는 게 더 안전하고 효율적인 방법이라고 판단해도 괜찮을까요?'
- 이것은 뷰를 어떻게 구성하냐에 따르고 뷰의 생성이 다른 것에 대한 의존성이 있는 경우엔 옵셔널 또는 lazy var를 사용하기도 합니다. 다만 옵셔널 처리를 하더라도 저는 강제옵셔널에 대한 부분은 조심해야하지 않을까했어요~ (+ 옵셔널로 표기되었을 때 의미하는 바가 있을겁니다.)

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        setupCollectionView()
+        setupActivityIndicator()
+        registerCell()
+        autoCheckNewProduct(timeInterval: 10)
```
여기에서만 쓰이고 고정된값을 쓴다면 autoCheckNewProduct메서드 내부에서 timeInterval을 쓰는것과 이렇게 외부에서 주는것과 어떤 차이가 있나요?

hwangjeha replied:
고정된 값을 함수 내부에서 사용한다면 함수 호출 시 고정된 값을 수정할 수 없게 됩니다.
매개변수 기본 값을 사용하면 함수 호출 시 timeInterval을 수정도 할 수 있고, `autoCheckNewProduct()` 처럼 사용할 수 있는 장점 있을 것 같습니다 :)

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductAlertPresenter.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit
+
+// MARK: - AddProduct Alert Presenter
+extension AddProductViewController {
```
AlertPresenter라는 네이밍이 현재 이 파일의 구현된 사항들과 쉽게 떠오르거나 매칭되는거 같진 않습니다 🥲

just1103 replied:
`Register Result Alert`로 네이밍을 수정했고, Utility 폴더에 `AlertFactory`타입을 생성하여 Alert 및 ActionSheet 생성을 처리하도록 개선했습니다!

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductListViewController/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    }
+    
+    private func setupCollectionView() {
+        productCollectionView = productListStackView.productCollectionView
```
스택뷰 내부에서 컬렉션뷰를 생성하고, 해당 컬렉션뷰를 이 화면의 컬렉션뷰로 사용하시는군요
뷰컨에서 직접 스택뷰와 컬렉션뷰를 가지고 있지 않고 이렇게 스택뷰가 가지고 있고, 뷰컨에서 그것을 참조하도록 하는 이유는 무엇인가용? 뷰컨에서 스택뷰와 컬렉션뷰를 가지고 있을때 우려되는 점이 있으신걸까요?

hwangjeha replied:
새로고침 버튼이 추가되면서 첫번째 화면의 전체 뷰는 스택뷰가 담당하고 있습니다.
현재 프로젝트에서 뷰를 관리하는 방식은 상위 뷰가 하위 뷰를 담고 있는 방식을 취하고 있습니다.
컬렉션 뷰를 뷰컨에서 생성하고 스택뷰에 addArrangedSubview 해주는 방식으로 변경할 수 있지만
그렇게 되면 이 화면에 대한 뷰 구성을 개발자가 기억해야하는 부담이 있을 것 같습니다.

추가적으로 현재 구조에서는 스택 뷰도 컬렉션 뷰를 참조하고 있고, 뷰컨도 컬렉션 뷰를 참조하고 있어
컬렉션 뷰의 RC가 이중으로 올라갈 수 있다고 생각되어 뷰컨의 컬렉션 뷰 프로퍼티를 약한 참조(weak)를 사용해여 개선했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductDataSource.swift:
```diff
@@ -0,0 +1,159 @@
+import UIKit
+
+private extension UIView {
+    func findSuperview<T>(ofType: T.Type) -> T? {
```
제네릭의 활용 멋집니다 👍

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductDataSource.swift:
```diff
@@ -0,0 +1,159 @@
+                                                      height: UIScreen.main.bounds.width * 0.35)
+    var imageAddFooterViewSize = AddProductDataSource.defaultImageAddFooterViewSize
+    
+    var productImages: [UIImage] = [] {
```
Good Naming!

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductDataSource.swift:
```diff
@@ -0,0 +1,159 @@
+            return
+        }
+        
+        dataSource.productImages.remove(at: indexPath.item)
```
이 부분도 지금은 AddProductViewController가 productImages에 직접 접근해서 remove라는 일을 하고 있는데요. DataSource한테 메서드로 행위에 필요한것들을 넘겨주고 내부에서 이런 작업들이 진행되도록 하는건 어떤가요? 
만약에 요구사항의 변경으로 productImages에 대한 행동이 바뀌어야한다면 여기 코드 변경없이 실제 productImages를 가지고 있는 dataSource 내부코드 변경만 이루어지도록 할 수 있지 않을까요

hwangjeha replied:
두 번째 화면 (상품 등록)에도 델리게이트 패턴 적용해서 역할 분리를 해줬습니다:)

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductDataSource.swift:
```diff
@@ -0,0 +1,159 @@
+        case .success(let generatedData):
+            return generatedData
+        case .failure(.invalidNameCount):
+            showRegisterFailAlert(message: .nameFailMessage)
```
여기 아래 코드들의 중복을 줄일 수는 없을까요?

just1103 replied:
GenerateUserInputError 열거형의 description 프로퍼티를 추가하여 아래 코드처럼 개선했습니다. 👍 
```swift
  case .failure(let error):
      showRegisterFailAlert(message: error.description)
```
이때 description을 String 타입이 아니라 `ProductRegisterAlertText` 타입으로 지정하여 선택지를 제한하도록 했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductImagePicker.swift:
```diff
@@ -0,0 +1,97 @@
+import UIKit
+
+// MARK: - ImagePicker Methods
```
이 파일명도 AddProductImagePicker 타입을 기대하게 되네요 🥲

just1103 replied:
역할 분리를 위해 ProductRegisterImagePicker 타입을 추가하고, delegate 패턴을 적용하여 개선했습니다!

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductViewController.swift:
```diff
@@ -0,0 +1,139 @@
+enum ViewTitle: String {
+    case productRegister = "상품등록"
+    
+    fileprivate var description: String {
```
어차피 여기서 description을 정의해서 쓸거라면 rawValue를 쓰기보단 스위치문을 쓰는건 어떤가요?

hwangjeha replied:
```swift
enum ViewTitle: String {
    case productRegister = "상품등록"
    case productModify = "상품수정"
    
    fileprivate var description: String {
        return self.rawValue
    }
}
```
처럼 사용하려는 의도로 작성했습니다.

STEP 1, 2에서도 위와 같은 방식을 사용해서 통일감을 주려고 사용했습니다.
저희가 생각해본 바로는 case의 rawValue가 수정되면 description을 사용한 모든 곳에 사이드이팩트가 일어나게 될 것으로 예상됩니다.
혹시 rawValue를 사용함에 있어 또 다른 문제가 발생할 수 있을까요??

myssun0325 replied:
지금 코드는 `ViewTitle.productRegister.rawValue` `ViewTitle.productRegister.description`이 같다는 얘기였습니다. 같은 값을 굳이 변수를 하나 더 추가해서 사용해야할까요?

스위치문을 쓴다는것은 아래와 같은 코드를 말씀드린거였습니당 :)

```swift
var description: String {
    switch self {
    case .productRegister:
        return "상품등록"
    case .productModify:
        return "상품수정"
    }
}
```

myssun0325 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController/AddProductViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    private(set) var productManagementScrollView = ProductManagementScrollView()
+    private(set) var imageCollectionView: ProductImageCollectionView!
+    private var descriptionTextView: UITextView!
+    private(set) var imagePickerController = UIImagePickerController()
```
이 부분도 자꾸 파일분리로 인해서 열게되네요 ㅠ
추가적으로 가져가야하는 단점중에 하나라고 생각하는 것은 내부에서만 쓰이는 값인데도 이렇게 열리면 코드를 트래킹하기 굉장히 어렵다는겁니다. 변경을 예측할 수가 없게 되고, 이 부분에 대해서 내부적인 코드를 짤때에 다른 파일의 변경사항들까지도 고려를 해야되는거죠

myssun0325 commented on OpenMarket/OpenMarket/Extension/UIImage+Extension.swift:
```diff
@@ -0,0 +1,46 @@
+        let compressedSize = self.calculateImageResizingRatio()
+        let resizedImage: UIImage?
+        
+        if #available(iOS 15.0, *) {
```
👍

myssun0325 commented on OpenMarket/OpenMarket/Model/ProductPlaceholder.swift:
```diff
@@ -0,0 +1,13 @@
+    case stock = "재고수량"
+    case description = "제품의 설명을 작성해주세요.\n글자수는 1000자 미만까지 가능합니다.\n\n\n\n\n"
+    
+    var text: String {
```
여기도 동일합니다. 외부에서 rawValue로 접근하는것과의 차이는 네이밍정도인듯합니당

just1103 replied:
윗부분 코멘트 다시 달아주시면 일괄수정하겠습니다. :)

myssun0325 commented on OpenMarket/OpenMarket/Model/UserInputChecker.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+    
+    func checkDescription(_ description: String) -> Bool {
+        if description.count >= 10,
```
조건식 자체를 반환값으로 쓸수 있도록 할수도 있겠군요

hwangjeha replied:
다른 메서드들과 동일하게 조건문을 바로 return 하도록 수정했습니다.
또, 수량이 음수가 되는 경우를 막기 위해서 checkStock()를 추가했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Model/UserInputChecker.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+struct UserInputChecker {
```
이렇게 내부에 프로퍼티 없이 메서드만 있는 타입을 다룰 때 메서드가 static인것과 아닌것은 어떤 차이가 있고 어떤 이점이 있을까요

just1103 replied:
`static`을 사용하면 구조체 인스턴스를 만들 필요가 없어서 호출 시 편리하다는 장점이 있지만, 메모리에 항상 올라가있다는 단점이 있습니다.

myssun0325 commented on OpenMarket/OpenMarket/Protocol/APIProtocol.swift:
```diff
@@ -5,13 +5,24 @@ protocol APIProtocol {
+
 enum HttpMethod: CustomStringConvertible {
     case get
+    case post
```
Good
확장에 용이하군요

myssun0325 commented on OpenMarket/OpenMarket/Utility/MultipartFormData.swift:
```diff
@@ -0,0 +1,121 @@
+    
+    init() {
+        let uuid = UUID().uuidString
+        self.boundary = "Boundary-\(uuid)"
```
init과 처음 default값으로 주는것과 어떤 기준을 가지고 init에서 주게되었나요?

hwangjeha replied:
UUID를 init 내부에서 생성할 지, 매개변수 기본값으로 UUID를 생성해서 넘겨줄지에 대한 기준은 따로 없었습니다.
UUID로 임의의 boundary를 생성하기 때문에 내부에서 생성해도 괜찮겠다는 생각이 였습니다.

피드백 주신걸 토대로 회의를 진행했는데 확장성을 고려한다면 매개변수 기본값으로 UUID를 넘겨주는게 더 용의하겠다는 판단이 들어  `init(uuid: String = UUID().uuidString)`로 개선했습니다.

만약, 앱 사용자마다 별도의 UUID를 가지고 서버에서 사용자를 특정하는 기능이 추가될 경우 매개변수의 기본값으로 UUID를 생성해서 넘겨주는것이 좋겠다는 생각을 해봤습니다.

myssun0325 commented on OpenMarket/OpenMarket/View/ProductListView/ProductsCollectionView.swift:
```diff
@@ -0,0 +1,38 @@
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        if let handler = reloadDataCompletionHandler {
```
아래 코드들을 layoutSubviews에 넣은 이유가 뭔가요? 
layoutSubviews는 언제 불리는 메서드인가요?

just1103 replied:
reloadData가 완료되는 시점에 layoutSubviews 메서드를 호출하고, completion을 실행하기 위해 위와 같이 작성했습니다. reloadData가 호출되면 CollectionView의 dataSource 및 delegate 메서드, 그리고 layoutSubviews 메서드가 차례로 호출된다는 것을 이용했습니다. 
(layoutSubviews 메서드는 하위 뷰의 layoutSubviews 메서드를 연쇄적으로 호출한다고 파악했습니다.)

이 부분에 대해 STEP2 README [2. 화면 전환 시 애니메이션 버그](https://github.com/just1103/ios-open-market/tree/step2#2-4-Trouble-Shooting)에 자세히 정리해두었습니다.

myssun0325 reviewed (commented):
수고하셨습니다 쀼의 세계에 오신걸 환영합니다 👏👏👏
객체가 커짐에 따라 분리를 하거나 이렇게 많은 뷰를 코드로 작성하셨는데도 깨지는 오류없이 잘 구현해내신거 같습니다 :) 

전체적으로 파일의 구성이나 extension의 기준 때문에 타협하게 되는 프로퍼티들이 보입니다. 코드와 파일트리 중 어느것을 우선순위로 기준을 잡고 피벗하면서 코드나 파일트리를 건드려야할지 고민하고 개선할 수 있을듯합니다. 

아래 코멘트와 함께 코드에 남겨진 리뷰를 확인해보세요!


1. Tap Gesture
- 구현내용 5번과도 연관이 있는거 같군요 이 부분에 대해서는 Responder Chain이 어떻게 동작하는지를 조금 더 학습해보셔야할거같습니다. 원인을 유추했다면 실질적으로 샘플코드 작성을 통해 확인을 해보는게 좋을거 같습니다. 어떻게 유추한것을 실제로 확인을 할 수 있을지도 고민해봐야하는 포인트라고 생각합니다. 현업에서도 요구사항과 뷰를 고려한 UX에 따라 다르게 처리하고 있습니다. 실질적으로 셀안에 버튼을 넣는 경우도 있고, 제스터인식기를 다는 경우도 있습니다. 자연스러운 시나리오들이라고 생각합니다. 다만 꼭! 어떻게 동작하는지를 명확히 짚고 넘어가셨으면 합니다 :) 
2. ViewController 거대화
- 일단 한 객체의 역할이 너무 많아지고, 가지고 역할과 책임을 나눠주기 위해 DataSource라는 객체를 선택하신건 굉장히 좋은 방법 같습니다!
제가 네이밍이나 예상한거랑은 조금 다르게 DataSource객체를 생성했지만, 컬렉션뷰의 데이터소스 역할은 그대로 OpenMarketViewController가 담당하고 있는듯 합니다.(extension으로 눈에 보이는 코드를 분리했을 뿐이지, 실질적으로 뷰컨트롤러가 책임을 지고 있는 모양이네요) OpenMarketDataSource객체가 실질적으로 그 역할을 가져가도록 개선할 수 있을까요?
4. CollectionView의 deleteItem 메서드
- 당연히 이런 경우에 사용하는 메서드입니다. (이런 경우와 애니메이션이 전체적으로 필요할 때 performBatchUpdates와 함께 사용하곤합니다.)
- 현재 수정 이전버전에서 문제가 되는 것은 뷰와 실제 데이터의 동기화의 문제인듯합니다. 컬렉션뷰에서 delete를 하고나서 reload되는 시점에서 실제 데이터와 일치하지 않는다는 얘기입니다. 요 부분에 대해서는 브레이크포인트를 찍어보시것과 수정 이전 버전에서 productImages의 remove를 해보시면 충분히 버그를 잡으실 수 있을거라고 생각합니다. (단 이 때, 한가지 더 고려해야하는것이 있는데, 2개이상의 사진을 추가하고 하나씩 지워갈때 indexPath에 신경을 쓰셔야합니다.)
5. 이 부분에 대해선 궁금한 기준이 있습니다~! 어떤 단위로 뷰컨에 직접 뷰를 그리기보다 커스텀뷰 내부에 뷰들을 쌓아서 올리는지 그 기준이 궁금합니당. 답변을 토대로 대화를 이어나가봅시다! 사실 뷰는 굉장히 이렇다할 정답이 있는것은 아니고 로직보다 구현할 방법이 훨씬 많고 눈에 보여지는것이라 구성하는 방법도 굉장히 다양하거든요. 대신 기능적으로 우려되는 위험사항들에 대해선 함께 논의 해볼 수 있을듯 합니다.!

just1103 committed:
refactor: 상수 cellHeightRatio/cellWidthRatio 추가 #7

- 하드코딩 최소화함

Commit:
refactor: OpenMarketViewController 옵셔널 프로퍼티 개선 #7

- LayoutKindSegmentedControl 이니셜라이저 개선

just1103 committed:
refactor: AlertFactory 타입 추가 #7

- Alert 및 ActionSheet 생성을 처리하는 createAlert 메서드 추가

Commit:
refactor: productCollectionView 약한 참조 사용 #7

- ProductListStackView, OpenMarketViewController가 이중으로 참조하고 있어
  OpenMarketViewController의 프로퍼티에 weak를 사용함.

just1103 committed:
refactor: GenerateUserInputError 열거형의 description 프로퍼티 추가 #7

Commit:
refactor: UserInputChecker 개선 #7

- checkDescription, checkName 조건식 수정
- checkStock 추가

just1103 committed:
refactor: MultipartFormData 이니셜라이저 개선 #7

- 이니셜라이저 기본값 지정
- ListProductCell pricesStackView의 distribution 변경

Commit:
refactor OpenMarketViewController Delegate 패턴 적용 #7

just1103 committed:
refactor: ProductListLayout 타입 추가 #7

- OpenMarketViewController의 Layout 역할을 분리함

Commit:
style: 파일명 수정 (OpenMarket -> ProductList) #7

just1103 committed:
refactor: AddProductViewController Delegate 패턴 적용 #7

- AddProductDataSourceDelegate 프로토콜 추가
- AddProductDataSource의 delegate를 AddProductViewController로 설정

Commit:
refactor: 상품 이미지 수정/삭제 Delegate Pattern 적용 #7

just1103 committed:
refactor: ProductRegisterImagePicker 타입 추가 및 Delegate 패턴 적용 #7

- ProductRegisterImagePickerDelegate 프로토콜 추가
- 파일명 변경 (AddProductViewController -> ProductRegisterViewController 등)

Commit:
refactor: 상품 등록(Post) 델리게이트 패턴 적용 #7

- 가격 확인 로직 개선(메세지 변경)
